src/t8_geometry/t8_geometry_helpers.c

```diff
@@ -201,3 +220,223 @@ t8_geom_get_edge_vertices (const t8_eclass_t tree_class,
+      ref_slope = 0;
+    }
+    /* slope = (y2-y1)/(x2-x1) */
+    else {
+      ref_slope = (ref_coords[1] - ref_opposite_vertex[1])
+        / (ref_coords[0] - ref_opposite_vertex[0]);
+    }
+    break;
+  case 2:                      /* edge 2 */
+    /* opposite vertex -> vertex 2 */
+    ref_opposite_vertex[0] = 1;
+    ref_opposite_vertex[1] = 1;
+    /* In case the reference point is equal to the opposite vertex, the slope of the line is 0. */
+    if (ref_coords[0] == ref_opposite_vertex[0]) {
+      ref_slope = 0;
+    }
+    /* slope = (y2-y1)/(x2-x1) */
+    else {
+      ref_slope = (ref_coords[1] - ref_opposite_vertex[1])
+        / (ref_coords[0] - ref_opposite_vertex[0]);
+    }
+    break;
+  default:
+    SC_ABORT_NOT_REACHED ();
+    break;
+  }
+  /* Now that we have the slope of the lines going through each vertex and the reference point,
+   * we can calculate the intersection of the lines with each edge.
+   */
+  switch (edge_index) {
+  case 0:                      /* edge 0 */
+    /* Because of the verticality of edge 0, the x value of the intersection is always 1.
+     * The y value is determined by the slope times the horizontal edge length.
+     */
+    ref_intersection[0] = 1;
+    ref_intersection[1] = ref_slope * 1;
+    break;
+  case 1:                      /* edge 1 */
+    /* If the reference point lies somewhere on edge 0, the intersection has to be at (1,1). */
+    if (ref_coords[0] == ref_opposite_vertex[0]) {
+      ref_intersection[0] = 1;
+      ref_intersection[1] = 1;
+      break;
+    }
+    /* If the reference point lies somewhere on edge 2, the intersection has to be at (0,0). */
+    else if (ref_coords[1] == ref_opposite_vertex[1]) {
+      ref_intersection[0] = 0;
+      ref_intersection[1] = 0;
+      break;
+    }
+    else {
+      /* intersectionX = (x1y2-y1x2)(x3-x4)-(x1-x2)(x3y4-y3x4)
+       *                 /(x1-x2)(y3-y4)-(y1-y2)(x3-x4)
+       * intersectionY = (x1y2-y1x2)(y3-y4)-(y1-y2)(x3y4-y3x4)
+       *                 /(x1-x2)(y3-y4)-(y1-y2)(x3-x4)
+       *
+       * c1 is the y axis intercept of the hypotenuse
+       * m1 is the slope of the hypotenuse
+       */
+      // x1=0 y1=0 x2=1 y2=1 x3=ref_coords[0] y3=ref_coords[1] x4=ref_opposite_vertex[0] y4=ref_opposite_vertex[1]
+      ref_intersection[0] =
+        ((0 * 1 - 0 * 1) * (ref_coords[0] - ref_opposite_vertex[0]) -
+         (0 - 1) * (ref_coords[0] * ref_opposite_vertex[1] -
+                    ref_coords[1] * ref_opposite_vertex[0]))
+        / ((0 - 1) * (ref_coords[1] - ref_opposite_vertex[1]) -
+           (0 - 1) * (ref_coords[0] - ref_opposite_vertex[0]));
+      ref_intersection[1] =
+        ((0 * 1 - 0 * 1) * (ref_coords[1] - ref_opposite_vertex[1]) -
+         (0 - 1) * (ref_coords[0] * ref_opposite_vertex[1] -
+                    ref_coords[1] * ref_opposite_vertex[0]))
+        / ((0 - 1) * (ref_coords[1] - ref_opposite_vertex[1]) -
+           (0 - 1) * (ref_coords[0] - ref_opposite_vertex[0]));
```

Comment:
You can simplify these equations by hand, to use the least amount of operations possible.
Furthermore, bitshift operations are the fastest operations possible. Slightly slower are additions and subtractions and after that multiplications are even slower.
Use divisions, modulo, pow and sqrt as sparsely as possible, because they are very slow. Maybe you can factor out some parts of the equation, so that a minimal amount of these operations are used.